iceoryx_utils/include/iceoryx_utils/cxx/method_callback.hpp

```diff
@@ -0,0 +1,172 @@
+    ///        on a different class.
+    /// @param[in] classPtr class pointer
+    template <typename ClassType>
+    void setObjectPointer(ClassType* const classPtr) noexcept;
+
+    /// @brief Returns classPtr
+    template <typename ClassType>
+    ClassType* getClassPointer() const noexcept;
+
+  private:
+    void* m_classPtr{nullptr};
+    ReturnValue (internal::GenericClass::*m_methodPtr)(Args...) const;
+    cxx::function_ref<ReturnValue(void*, ReturnValue (internal::GenericClass::*)(Args...) const, Args...)> m_callback;
+};
+
+template <typename ReturnValue, typename... Args>
+class MethodCallback
+{
+  public:
+    MethodCallback() = default;
+    MethodCallback(const MethodCallback& rhs) = default;
+    MethodCallback& operator=(const MethodCallback& rhs) = default;
+    ~MethodCallback() = default;
+
+    /// @brief Constructs a MethodCallback from a pointer to a specific class
+    ///        and a pointer to a method of that class.
+    /// @param[in] classPtr class pointer
+    /// @param[in] methodPtr pointer to a method
+    template <typename ClassType>
+    MethodCallback(ClassType* const classPtr, ReturnValue (ClassType::*methodPtr)(Args...)) noexcept;
+
+    /// @brief Move constructor
+    /// @param[in] rhs move origin
+    MethodCallback(MethodCallback&& rhs) noexcept;
+
+    /// @brief Move assignment operator
+    /// @param[in] rhs move origin
+    /// @return reference to this
+    MethodCallback& operator=(MethodCallback&& rhs) noexcept;
+
+    /// @brief Calls the method if the MethodCallback is valid, otherwise it
+    ///         will return MethodCallbackError::UNABLE_TO_CALL_METHOD_ON_NULLPTR_CLASS_PTR
+    /// @param[in] args... arguments which will be perfectly forwarded to the method
+    /// @return If MethodCallback is valid the return value of the method, otherwise
+    ///         an error.
+    template <typename... MethodArguments>
+    expected<ReturnValue, MethodCallbackError> operator()(MethodArguments&&... args) noexcept;
+
+    /// @brief Comparison operator. Two MethodCallbacks are equal if they have the
+    ///        same class pointer and method pointer
+    bool operator==(const MethodCallback& rhs) const noexcept;
+
+    /// @brief Inequality operator. Two MethodCallbacks are not equal if they have
+    ///        different class or method pointer
+    bool operator!=(const MethodCallback& rhs) const noexcept;
+
+    /// @brief Verifies if the MethodCallback is valid.
+    /// @return true if classPtr != nullptr otherwise false
+    explicit operator bool() const noexcept;
+
+    /// @brief Verifies if the MethodCallback is valid.
+    /// @return true if classPtr != nullptr otherwise false
+    bool isValid() const noexcept;
+
+    /// @brief Sets a new classPtr. The callback will then call the same method
+    ///        on a different class.
+    /// @param[in] classPtr class pointer
+    template <typename ClassType>
+    void setObjectPointer(ClassType* const classPtr) noexcept;
```

Comment:
I think we need to properly distinguish between classes and objects of classes. We set the object pointer here. The getter should have a corresponding name and the comment should reflect this as well.

Reply:
renamed the method to `getObjectPointer`